clients/src/main/java/org/apache/kafka/clients/consumer/internals/DefaultBackgroundThreadRunnable.java

```diff
@@ -0,0 +1,235 @@
+    DefaultBackgroundThreadRunnable(Time time,
+                                    ConsumerConfig config,
+                                    BlockingQueue<ApplicationEvent> applicationEventQueue,
+                                    BlockingQueue<BackgroundEvent> backgroundEventQueue,
+                                    SubscriptionState subscriptions,
+                                    ConsumerMetadata metadata,
+                                    ConsumerNetworkClient client) {
+        this.time = time;
+        this.config = config;
+        this.log = LoggerFactory.getLogger(getClass());
+        this.applicationEventQueue = applicationEventQueue;
+        this.backgroundEventQueue = backgroundEventQueue;
+        this.subscriptions = subscriptions;
+        this.metadata = metadata;
+        this.networkClient = client;
+        this.metrics = new Metrics();
+    }
+
+    private void setConfig() {
+        this.retryBackoffMs = this.config.getLong(ConsumerConfig.RETRY_BACKOFF_MS_CONFIG);
+        this.clientId = config.getString(CommonClientConfigs.CLIENT_ID_CONFIG);
+        this.heartbeatIntervalMs = config.getInt(ConsumerConfig.HEARTBEAT_INTERVAL_MS_CONFIG);
+    }
+
+    @Override
+    public void run() {
+        try {
+            log.debug("DefaultBackgroundThreadRunnable started");
+            while (running) {
+                pollOnce();
+                this.wait(retryBackoffMs);
+            }
+        } catch (Exception e) {
+            // TODO: Define fine grain exceptions here
+        } finally {
+            log.debug("DefaultBackgroundThreadRunnable closed");
+        }
+    }
+
+    /**
+     * Process event from a single poll
+     */
+    void pollOnce() {
+        this.inflightEvent = maybePollEvent();
+        if (this.inflightEvent.isPresent() && maybeConsumeInflightEvent(this.inflightEvent.get())) {
+            // clear inflight event upon successful consumption
+            this.inflightEvent = Optional.empty();
+        }
+        networkClient.pollNoWakeup();
+    }
+
+    public Optional<ApplicationEvent> maybePollEvent() {
+        if (this.inflightEvent.isPresent()) {
+            return this.inflightEvent;
+        }
+        return Optional.ofNullable(this.applicationEventQueue.poll());
+    }
+
+    /**
+     * ApplicationEvent are consumed here.
+     * @param event an {@link ApplicationEvent}
+     * @return true when successfully consumed the event.
+     */
+    public boolean maybeConsumeInflightEvent(ApplicationEvent event) {
+        switch (event.type) {
+            case NOOP:
+                log.info("consuming a NOOP event");
+                return true;
+            default:
+                inflightEvent = Optional.empty();
+                log.info("unsupported event type: {}", event.type);
+        }
+        return false;
+    }
+
+    @Override
+    public void close() {
+        this.running = false;
+        Utils.closeQuietly(networkClient, "consumer network client");
```

Comment:
I believe the `Metrics` and `ConsumerMetadata` classes have a `close()` method that should be invoked here too.

Reply:
I wonder if the metrics should be closed by the polling thread as it is also used there.